packages/cli/src/commands/auth/login/loginFlow.ts

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2023 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// eslint-disable-next-line import/no-named-as-default
+import consola from "consola";
+import { createServer } from "http";
+import open from "open";
+import { join } from "path/posix";
+import { parse } from "url";
+import type { LoginArgs } from "./LoginArgs.js";
+
+export default async function invokeLoginFlow(args: LoginArgs) {
+  consola.start(`Authenticating using application id: ${args.applicationId}`);
+  const redirectUrl = "http://localhost:8080/auth/callback";
+  const port = parse(redirectUrl).port;
+  let resolve: (value: string) => void;
+  const authCode: Promise<string> = new Promise((_res) => {
+    resolve = _res;
+  });
+
+  const server = createServer((req, res) => {
+    const query = parse(req.url!, true).query;
+    res.end("Authenticated");
+    resolve(query["code"] as string);
+  });
+
+  server.listen(port);
+  const clientId = args.applicationId;
+
+  const state = generateRandomString();
+  const codeVerifier = generateRandomString();
+  const codeChallenge = await generateCodeChallenge(codeVerifier);
+
+  const authorizeUrl = generateAuthorizeUrl(
+    args.baseUrl,
+    clientId,
+    state,
+    redirectUrl,
+    codeChallenge,
+  );
+
+  try {
+    await open(authorizeUrl);
+  } catch {
+    consola.warn(
+      `Unable to open browser, please open this url in your browser to to authenticate: ${authorizeUrl}`,
+    );
+  }
+
+  const code = await authCode;
+  server.close();
+
+  const token = await getTokenWithCodeVerifier(
+    clientId,
+    redirectUrl,
+    code,
+    args.baseUrl,
+    codeVerifier,
+  );
+
+  consola.success(`Successfully authenticated!`);
+  return token;
+}
+
+function generateRandomString() {
+  const array = crypto.getRandomValues(new Uint32Array(28));
+  return Array.from(array, dec => {
+    const decimalString = "0" + dec.toString(16);
+    return decimalString.substring(decimalString.length - 2);
```

Comment:
Took me a sec to figure this out. Also is a length of 28 part of a spec somewhere?

```suggestion
    return dec.toString(16).padStart(2,"0");
```